webapp/src/components/expanded_view/call_settings.tsx

```diff
@@ -0,0 +1,477 @@
+`;
+
+type AudioDevicesProps = {
+    deviceType: string;
+    isActive: boolean;
+    onToggle: (deviceType: string) => void;
+}
+
+const AudioDevices = ({deviceType, isActive, onToggle}: AudioDevicesProps) => {
+    const {formatMessage} = useIntl();
+    const [currentAudioInputDevice, setCurrentAudioInputDevice] = useState<MediaDeviceInfo | null>(null);
+    const [currentAudioOutputDevice, setCurrentAudioOutputDevice] = useState<MediaDeviceInfo | null>(null);
+    const [audioDevices, setAudioDevices] = useState<AudioDevices>({inputs: [], outputs: []});
+
+    const isInput = deviceType === 'input';
+
+    const handleDeviceClick = (device: MediaDeviceInfo) => {
+        const callsClient = getCallsClient();
+
+        if (isInput) {
+            if (device !== currentAudioInputDevice) {
+                callsClient?.setAudioInputDevice(device);
+            }
+            setCurrentAudioInputDevice(device);
+        } else {
+            if (device !== currentAudioOutputDevice) {
+                callsClient?.setAudioOutputDevice(device);
+            }
+            setCurrentAudioOutputDevice(device);
+        }
+
+        onToggle(deviceType);
+    };
+
+    const handleDeviceChange = (devices: AudioDevices) => {
+        setAudioDevices(devices);
+    };
+
+    useEffect(() => {
+        const callsClient = getCallsClient();
+        if (!callsClient) {
+            // eslint-disable-next-line no-undefined
+            return undefined;
+        }
+
+        callsClient.on('devicechange', handleDeviceChange);
+        setAudioDevices(callsClient.getAudioDevices());
+
+        if (isInput) {
+            setCurrentAudioInputDevice(callsClient.currentAudioInputDevice);
+        } else {
+            setCurrentAudioOutputDevice(callsClient.currentAudioOutputDevice);
+        }
+
+        return () => {
+            callsClient.off('devicechange', handleDeviceChange);
+        };
+    }, []);
+
+    if (isInput && audioDevices.inputs.length === 0) {
+        return null;
+    }
+
+    if (!isInput && audioDevices.outputs.length === 0) {
+        return null;
+    }
+
+    const currentDevice = isInput ? currentAudioInputDevice : currentAudioOutputDevice;
+    const DeviceIcon = isInput ? UnmutedIcon : SpeakerIcon;
+    const label = currentDevice?.label || formatMessage({defaultMessage: 'Default'});
+
+    const devices = isInput ?
+        audioDevices.inputs?.filter((device) => device.deviceId && device.label) :
+        audioDevices.outputs?.filter((device) => device.deviceId && device.label);
+    const isDisabled = devices.length === 0;
+
+    return (
+        <>
+            {isActive &&
+            <AudioDevicesList
+                deviceType={deviceType}
+                devices={devices}
+                currentDevice={isInput ? currentAudioInputDevice : currentAudioOutputDevice}
+                onDeviceClick={handleDeviceClick}
+            />
+            }
+            <li
+                className='MenuItem'
+            >
```

Comment:
nit: this (and others in the file) can be one line and would look cleaner.  But not a big deal.